up_client_socket/rust/u_transport_socket.rs

```diff
@@ -0,0 +1,395 @@
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Response UAttributes {e:?}"),
+                        )
+                    })?;
+                match socket_clone.write_all(&umsg_serialized) {
+                    Ok(_) => Err(UStatus::ok()),
+                    Err(_) => Err(UStatus::fail_with_code(
+                        UCode::UNAVAILABLE,
+                        "Dispatcher communication issue",
+                    )),
+                }
+            }
+            UMessageType::UMESSAGE_TYPE_UNSPECIFIED => Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Wrong Message type in UAttributes",
+            )),
+            UMessageType::UMESSAGE_TYPE_NOTIFICATION => Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Wrong Message type in UAttributes",
+            )),
+        }
+    }
+
+    /// Receives a message from the transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `topic` - The topic to receive the message from.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if no message could be received. Possible reasons are that the topic does not exist
+    /// or that no message is available from the topic.
+    async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+        Err(UStatus::fail_with_code(
+            UCode::UNIMPLEMENTED,
+            "Not implemented",
+        ))
+    }
+
+    /// Registers a listener to be called for each message that is received on a given address.
+    ///
+    /// # Arguments
+    ///
+    /// * `address` - The (resolved) address to register the listener for.
+    /// * `listener` - The listener to invoke.
+    ///
+    /// # Returns
+    ///
+    /// An identifier that can be used for [unregistering the listener](Self::unregister_listener) again.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the listener could not be registered.
+    ///
+    ///
+    ///
+    ///
+    ///
+
+    async fn register_listener(
+        &self,
+        topic: UUri,
+        listener: Arc<dyn UListener>,
+    ) -> Result<(), UStatus> {
+        dbg!("register listner called !");
+        if topic.authority.is_some() && topic.entity.is_none() && topic.resource.is_none() {
+            // This is special UUri which means we need to register for all of Publish, Request, and Response
+            // RPC response
+            Err(UStatus::fail_with_code(
+                UCode::UNIMPLEMENTED,
+                "Not implemented",
+            ))
+        } else {
+            // Do the validation
+            UriValidator::validate(&topic)
+                .map_err(|_| UStatus::fail_with_code(UCode::INVALID_ARGUMENT, "Invalid topic"))?;
+
+            if UriValidator::is_rpc_response(&topic) {
+                self.listner_map
+                    .lock()
+                    .unwrap()
+                    .entry(topic.to_string())
```

Comment:
Can remove the .to_string() if storing `UUri`s

Reply:
corrected